### ISSUE TITLE
High Acc but relatively low in matrices

### ISSUE THREAD

**Issue by ZehuaLiuY** [ZehuaLiuY/Species-Classification#9]
从代码和日志现象来看，“**准确率高但精确率、召回率、F1、mAP都不高**”往往与以下几点原因相关，尤其是在多类别且类别分布不均衡的场景中最常见：

---

## 1. 类别不平衡导致 accuracy 虚高

- **准确率 (Accuracy)** 只要在出现频率最高的类别上预测得足够多，就容易看上去很高；但那些少量样本的类别要么几乎不被预测到、要么预测不准，结果会明显拉低 **precision / recall / f1 / mAP**。
- 解决思路：
  - 查看下训练集、验证集中各个类别的样本数分布。若某些类的占比非常大，其他类非常小，就可能导致模型倾向于“大类”来获取更高的 overall accuracy。
  - 考虑在训练时引入 **class weights** 或者在取样时使用 **过采样/欠采样**、数据增强等方法来平衡训练。

---

## 2. 评价指标的多/单标签混用或宏平均宏观 (macro) 导致分数偏低

- 你在验证阶段使用了：

  ```python
  # 计算 P/R/F1
  precision_score(..., average='macro')
  recall_score(..., average='macro')
  f1_score(..., average='macro')
  ...
  
  # 计算 mAP
  average_precision_score(..., average="macro")
  ```

  对于多类单标签任务，“macro” 会先分别计算每个类别的精确率、召回率，再做平均。若数据极不平衡，少数类的表现会极大拉低平均值，导致 macro-F1/mAP 较低。  

- 同时你把 `labels` 做 `one_hot` 后用 `average_precision_score` 来算 **mAP**，这里实际上是按“多标签”的方式去算了每个类别的AP再做平均。如果实际是 **单标签多分类**，而且某些类别的正例少，就会出现 AP 指标整体偏低的问题。  
- 解决思路：
  - 如果实际上是 “**单标签多分类**”，而想要计算类似“mean Average Precision (mAP)” 的概念，需要对每个类别视为“一二分类”来分别算 AP，再做平均，这与多标签的计算思路类似，但要确认你的数据和目标是否真的需要这样做。  
  - 如果需要与常规分类任务指标对齐，可以先多看看 **per-class precision/recall** 再决定是使用 **macro** 还是 **weighted** 或 **micro**。  
  - 若确实是多标签（一个box可能对应多个类别），就需要在模型输出阶段进行多热预测等处理，但你现在好像是“一次只给出单一类别” —— 要检查这部分是否逻辑对齐。

---

## 3. 单批次裁剪多个 box、批量大小和分布可能不均

- 代码中，你对同一张图的多个 bounding boxes 分别裁剪，再一次性堆叠到一个 batch 中做训练/验证。  
- 如果某些图片/某些类别下 bounding boxes 特别多（或特别少），也会造成训练分布不平衡；在验证阶段也会遇到类似的分布不均衡问题，进一步放大“head 类”和“tail 类”的差距。

---

## 4. 阈值或后处理策略

- 在输出时你直接做 `torch.max(outputs, dim=1)` 选出分值最高的类，而在计算 AP、mAP 时，使用 `softmax` 的概率去算 `average_precision_score`。  
- 若某些类别的 logit 分数离主流类别差距很大，可能阈值一刀切就几乎不预测那些类别，也会拉低 recall，进而影响 F1 和 mAP。  
- 你可以尝试查看一下 **每个类别** 的评分曲线（PR曲线）以及把 threshold 从 0.5 改成其他值，再观察 precision/recall 的变化（对多分类可能还要分类别地查看，或者把 logit 分出来后看各自的分布）。

---

## 5. 实际需求 vs. 代码默认设置

- 你在代码里将模型最后一层改为 `torch.nn.Linear(num_features, 46)`，表示做 **46 类单标签** 分类；但是在计算 mAP 时却使用了 “多标签” 风格的 `average_precision_score`，还把标签 one-hot 做了 `filtered_labels` 和 `filtered_probs`。  
- **如果你的任务真的是单标签多分类**，可以考虑以下方式之一：
  1. **精确率/召回率/F1**：继续用 `precision_score(all_val_labels, all_val_preds, average='macro')` 也行，但要认识到类别不平衡时，macro-F1 不会很高。
  2. **mAP**：如果是单标签多分类，并不总是使用 mAP 来做评估；通常多用于目标检测或多标签分类。若一定要算，可以把每个类别视为“该类别 vs. 其他类别”来算AP，然后对46个类别做平均，这才能和你one-hot的方式对齐，但需要确认数据量、类别分布等是否适合。
- 如果任务其实是 **多标签多分类**（每个box可同时有多个标签），需要把最后输出改成多标签方式（如 `sigmoid` + 多标签损失函数 BCEWithLogitsLoss）并保证评估指标的计算与之匹配，现在看起来你是 `CrossEntropyLoss` + `softmax` + `torch.max`，那就只能输出一个单一类了。

---

## 建议排查/优化思路

1. **检查数据分布**  
   - 看看 46 个类别当中，是否有 1~2 个类别样本极多，其他类别样本极少；或者大部分boxes其实只有少数若干类别；  
   - 在这种情况下，模型往往倾向于预测“大头”类别，提高 overall accuracy。

2. **查看混淆矩阵/分类报告**  
   - 通过 `sklearn.metrics.classification_report` (per-class) 或自己统计混淆矩阵，看看到底是哪些类别预测得好，哪些差；多数时候你会发现某些极少数类别的 recall 很低，从而拉低 macro-F1。  
   - 也可把 per-class AP 算出来（如果真的要算多分类下的“mAP”）。

3. **重新审视多标签/单标签的设定**  
   - 如果实际上是单标签，就可以先用最常见的指标：accuracy + macro-F1 或 weighted-F1 等，不一定必须去算 mAP；  
   - 如果实际需求是多标签，要改网络最后一层激活函数和损失函数、并在评估中做多标签的阈值处理。

4. **尝试不同平均方式**  
   - `average='macro'` 会平等地对待每个类；如果数据分布极端不平衡，macro 值会看起来很低。可以试试 `average='weighted'` 或 `average='micro'` 看区别。

5. **（若仍用 mAP）阈值及后处理**  
   - 你现在的 `softmax` 输出是用于多分类，这时算 average precision score 要么对每个类分别当二分类去算，否则意义就有点混淆；  
   - 真正的 mAP 通常是“对每个类算PR曲线，积分得到AP，然后对所有类算平均”，可以参考 [[PyTorch官方对象检测教程](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html)](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html) 或 [[pycocotools mAP](https://cocodataset.org/#detection-eval)](https://cocodataset.org/#detection-eval) 的实现方式。

---

### 总结

- **最常见主因**：数据集中有一个或少数大类，使得整体 accuracy “看似很高”，但对小类的精准预测很差，于是 macro precision/recall/F1 和 mAP 都不高。  
- **快速验证**：检查每个类别的样本数、做个混淆矩阵/分类报告，就能很快发现到底是不是大部分预测都落到那几个类别上。  
- **代码层面**：如果确实是单标签多分类，mAP 的计算方式要和分类场景对齐，不要混用多标签的做法；如果是多标签，模型最后一层和损失函数也要做相应的多标签处理。  

只要先确定清楚“任务本质是单标签还是多标签”“类别分布情况怎样”“所用指标适用于什么场景”，然后再对症下药（均衡数据、调整阈值、改进模型结构等）即可。祝调试顺利!
